compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Upgrade.hs

```diff
@@ -283,86 +307,156 @@ generateTemplateInstance env typeCon typeParams externPkgId =
-                                                                            ":" <>
-                                                                            moduleNameStr <>
-                                                                            ":" <>
-                                                                            (T.unpack $
-                                                                             T.intercalate
-                                                                                 "." $
-                                                                             LF.unTypeConName
-                                                                                 templName) <>
-                                                                            ":" <>
-                                                                            T.unpack
-                                                                                funName))))
-                                                      ]
-                                                , grhssLocalBinds =
-                                                      noLoc emptyLocalBinds
-                                                }
-                                      }
-                                ]
-                      , mg_origin = Generated
+
+-- | Generate a single choice instance for a given templateDT/choice
+generateChoiceInstance ::
+       Env
+    -> LF.PackageId
+    -> LF.DefDataType
+    -> LF.TemplateChoice
+    -> HsDecl GhcPs
+generateChoiceInstance env externPkgId templateDT choice =
+    InstD noExt $
+    ClsInstD
+        noExt
+        ClsInstDecl
+            { cid_ext = noExt
+            , cid_poly_ty =
+                  HsIB
+                      { hsib_ext = noExt
+                      , hsib_body = body
                       }
-            , fun_co_fn = WpHole
-            , fun_tick = []
+            , cid_binds = listToBag $ map classMethodStub methodMapping
+            , cid_sigs = []
+            , cid_tyfam_insts = []
+            , cid_datafam_insts = []
+            , cid_overlap_mode = Nothing
             }
+  where
+
+    app :: LHsType GhcPs -> LHsType GhcPs -> LHsType GhcPs
+    app t1 t2 = noLoc $ HsAppTy noExt t1 t2
+
+    body :: LHsType GhcPs =
+      choiceClass `app` arg1 `app` arg2 `app` arg3
+
+    choiceClass :: LHsType GhcPs =
+        noLoc $
+        HsTyVar noExt NotPromoted $
+        noLoc $
+        mkRdrQual (mkModuleName "DA.Internal.Template") $
+        mkOccName varName "Choice" :: LHsType GhcPs
+
+    arg1 :: LHsType GhcPs =
+      noLoc $ convType env lfTemplateType
+
+    arg2 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceType
+
+    arg3 :: LHsType GhcPs =
+      noLoc $ convType env lfChoiceReturnType
+
+    moduleNameStr = T.unpack $ LF.moduleNameString $ LF.moduleName $ envMod env
+    moduleName0 =
```

Comment:
Isn’t this just `LF.moduleName $ envMod mod`? It looks like you are first intercalating dots via `moduleNameString`, unpacking only to then split on dots and repack.

Reply:
simplified to avoid this